messaging/broker.go

```diff
@@ -1232,33 +1233,59 @@ type Message struct {
 }
 
 // WriteTo encodes and writes the message to a writer. Implements io.WriterTo.
-func (m *Message) WriteTo(w io.Writer) (n int64, err error) {
-	if n, err := w.Write(m.marshalHeader()); err != nil {
-		return int64(n), err
-	}
-	if n, err := w.Write(m.Data); err != nil {
-		return int64(MessageHeaderSize + n), err
+func (m *Message) WriteTo(w io.Writer) (int64, error) {
+	b, err := m.MarshalBinary()
+	if err != nil {
+		return 0, err
 	}
-	return int64(MessageHeaderSize + len(m.Data)), nil
+	n, err := w.Write(b)
+	return int64(n), err
 }
 
 // MarshalBinary returns a binary representation of the message.
 // This implements encoding.BinaryMarshaler. An error cannot be returned.
 func (m *Message) MarshalBinary() ([]byte, error) {
-	b := make([]byte, MessageHeaderSize+len(m.Data))
-	copy(b, m.marshalHeader())
-	copy(b[MessageHeaderSize:], m.Data)
+	// Encode message.
+	b := make([]byte, MessageChecksumSize+MessageHeaderSize+len(m.Data))
+	copy(b[MessageChecksumSize:], m.marshalHeader())
+	copy(b[MessageChecksumSize+MessageHeaderSize:], m.Data)
+
+	// Calculate and write the checksum.
+	h := fnv.New32a()
+	h.Write(b[MessageChecksumSize:])
+	binary.BigEndian.PutUint32(b, h.Sum32())
+
 	return b, nil
 }
 
 // UnmarshalBinary reads a message from a binary encoded slice.
 // This implements encoding.BinaryUnmarshaler.
 func (m *Message) UnmarshalBinary(b []byte) error {
-	m.unmarshalHeader(b)
-	if len(b[MessageHeaderSize:]) < len(m.Data) {
-		return fmt.Errorf("message data too short: %d < %d", len(b[MessageHeaderSize:]), len(m.Data))
+	// Read checksum.
+	if len(b) < MessageChecksumSize {
+		return fmt.Errorf("message checksum too short: %d < %d", len(b), MessageChecksumSize)
+	}
+	sum := binary.BigEndian.Uint32(b)
+
+	// Read header.
+	if len(b)-MessageChecksumSize < MessageHeaderSize {
+		return fmt.Errorf("message header too short: %d < %d", len(b)-MessageChecksumSize, MessageHeaderSize)
+	}
+	m.unmarshalHeader(b[MessageChecksumSize:])
+
+	// Read data.
+	if len(b)-MessageChecksumSize-MessageHeaderSize < len(m.Data) {
+		return fmt.Errorf("message data too short: %d < %d", len(b)-MessageChecksumSize-MessageHeaderSize, len(m.Data))
+	}
+	copy(m.Data, b[MessageChecksumSize+MessageHeaderSize:])
+
+	// Verify checksum.
+	h := fnv.New32a()
+	h.Write(b[MessageChecksumSize:])
```

Comment:
The `Write()` functions in `hash/fnv` can't error so I ignore error checks on them:

http://golang.org/src/hash/fnv/fnv.go